src/workerd/io/worker.c++

```diff
@@ -448,37 +448,105 @@ public:
       // We're on a request-serving thread.
       auto& ioContext = IoContext::current();
       timePoint = ioContext.now();
-    } else KJ_IF_MAYBE(info, inspectorTimerInfo) {
-      if (info->threadId == getCurrentThreadId()) {
-        // We're on an inspector-serving thread.
+    } else {
+      auto lockedState = state.lockExclusive();
+      KJ_IF_MAYBE(info, lockedState->inspectorTimerInfo) {
         timePoint = info->timer.now() + info->timerOffset
                   - kj::origin<kj::TimePoint>() + kj::UNIX_EPOCH;
       }
+      // We're at script startup time -- just return the Epoch.
     }
+    return (timePoint - kj::UNIX_EPOCH) / kj::MILLISECONDS;
+  }
 
-    // If we're on neither a request- nor inspector-serving thread, then we're at script startup
-    // time -- just return the Epoch.
+  void setInspectorTimerInfo(kj::Timer& timer, kj::Duration timerOffset) {
+    auto lockedState = state.lockExclusive();
+    lockedState->inspectorTimerInfo = InspectorTimerInfo { timer, timerOffset, getCurrentThreadId() };
+  }
 
-    return (timePoint - kj::UNIX_EPOCH) / kj::MILLISECONDS;
+  void setChannel(Worker::Isolate::InspectorChannelImpl& channel) {
+    auto lockedState = state.lockExclusive();
+    lockedState->channel = channel;
   }
 
-  // Nothing else. We ignore everything the inspector tells us, because we only care about the
-  // devtools inspector protocol, which is handled separately.
+  void resetChannel() {
+    auto lockedState = state.lockExclusive();
+    lockedState->channel = {};
+  }
 
-  void setInspectorTimerInfo(kj::Timer& timer, kj::Duration timerOffset) {
-    // Helper for attachInspector().
-    inspectorTimerInfo = InspectorTimerInfo { timer, timerOffset, getCurrentThreadId() };
+  void runMessageLoopOnPause(int contextGroupId) override {
+    // This method is called by v8 when a breakpoint or debugger statement is hit. This method
+    // processes debugger messages until `Debugger.resume()` is called, when v8 then calls
+    // `quitMessageLoopOnPause()`.
+    //
+    // This method is ultimately called from the `InspectorChannelImpl` and the isolate lock is
+    // held when this method is called.
+    if (isMultiTenantProcess()) {
+      // TODO: breakpoints are not supported for the multi-tenant case.
+      return;
+    }
+
+    {
+      auto lockedState = state.lockExclusive();
+      KJ_IF_MAYBE(channel, lockedState->channel) {
+        pauseIncomingMessages(*channel);
+      }
+    }
+
+    runMessageLoop = true;
+    do {
+      auto lockedState = state.lockExclusive();
```

Comment:
It looks like this lock is held while we wait for the next message to arrive, and then is dropped only momentarily at the end of the loop and taken back a moment later.

If any other threads are actually contending for the lock this seems like a problem. Probably those other threads won't actually wake up fast enough to take the lock before the paused thread grabs it again. And if no subsequent message happens to be delivered, those threads could be waiting indefinitely.

If this isn't a problem could you expand the comments here to explain how so?

Reply:
So from the comment it sounds like this only matters on disconnect/shutdown? But won't shutdown cause dispatchOneMessageDuringPause() to return false, causing the loop here to break entirely? I still don't see why we need to release the lock for each iteration of the loop, rather than just at the end of the loop.

On the other hand though, I really feel like as a rule we shouldn't be waiting for I/O while holding a mutex lock. Is there a way we could structure this code so that it doesn't hold the lock while waiting for messages?

Reply:
I think you are right about the shutdown path. The last push fixes that.

I may be able to see a path to avoid the mutex...right now it is preventing deletion of the channel whilst processing messages, but if it is more closely associated with the WebSocketIoHandler this might be avoided.